### ISSUE TITLE
Investigate how hard would it be to support semicolons properly

### ISSUE THREAD

**Issue by mrkkrp** [tweag/ormolu#258]
By "support" I mean working with inputs that contain semicolons and outputting normal Ormolu output where semicolons are dropped.

There are some issues related to semicolons: https://github.com/tweag/ormolu/issues?q=is%3Aissue+is%3Aopen+label%3Asemicolons.

The aim of this issue is to investigate:

1. Where semicolons can appear in Haskell syntax in general, for that Haskell report may be useful.
2. Try to assess how hard would it be to support semicolons with what we currently have.
3. What sort of problems prevent us from supporting semicolons.

It may be that semicolons are too tricky to support properly. In that case we'll proceed with #253, otherwise we're going to implement a solution to support semicolons and fix the related bugs.

**Comment by mrkkrp** [comment]
Some packages that currently fail because of semicolons (may be useful for checking real-world examples):

* `aeson`
* `lens`
* `aws`
* `parsec3`

**Comment by utdemir** [comment]
The first thing I can think of is that, converting semicolon syntax to a multi line construct breaks idempotency in the current design. Consider this example

```
let a = a in do { foo; bar; }
```

On the first run, ormolu would think that it is a single line construct, and start formatting it in that way, however it would split that `do` notation to multiple lines when printing:

```
let a = a in do
  foo
  bar
```

But on the next run, it would notice that it's a multiline construct and format it accordingly:

```
let a = a
 in do
      foo
      bar
```

As you can see, somehow we have to detect that our layout is going to be a multiline layout before we even hit the `do`. To achieve this, we will probably have to have a preprocessing step where we convert semicolon-syntax to identation-syntax, and format the result. Or as a more hacky solution, we can simply format the code once more if we detect this issue.

**Comment by utdemir** [comment]
As per Haskell 2010 report, we are be able to use semicolons everywhere layout matters, including the top-level. Below is a type-correct Haskell program with non exhaustive set of examples:

```haskell
{-# LANGUAGE GADTs #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TemplateHaskell #-}

module Semicolons where
{ import Prelude
; f0 = let { a = 1; b = 2 } in b
; f1 = let a = 1; b = 2
           c = 3
       in  b
; f2 = a where { a = 3 }
; f3 = do a <- [1]; let { b = a }; return b
; f4 = case True of { True -> 1; False -> 2 }
; f5 = case f2  of 1 -> 1; 2 -> 2
                   3 -> 3
; data Foo a where Bar :: Foo (); Baz :: Foo Int
; class Cls a where type T a; get :: a -> T a
; instance Cls (Foo a) where type T (Foo a) = Int; get = undefined
; type family G a where G Int = Bool; G a = Char
; main = print (f0, f1, f2, f3, f4, f5)
; do return 1; return []
}
```

As far as I understand, every Haskell source file (ignoring `-XQuasiQuotes`) can be transformed to a single line using braces and semicolons.

**Comment by utdemir** [comment]
In light of those, here is what I think we can do to support those:

* Convert semicolon-syntax to indentation-syntax: This would require either inspecting and modifying the AST before formatting or formatting twice.
* Preserve semicolons: This requires some refactoring, but I do not think we'll end up with twice as much code, since the way semicolon-syntax and the indentation-syntax handled is pretty uniform, so we should be able to create a combinator like `{ data LayoutMode = Semicolon | Indentation; printLayout :: LayoutMode -> [R ()] -> R () }`.
* Or, we can only support semicolon-syntax on single line constructs but I don't think there is any advantage doing that.

**Comment by mrkkrp** [comment]
> Of course we first have to figure out a way to check if an expression is using the semicolon syntax or not, it looks like GHC API does not provide an easy option.

FTR, the way you fan find out about semicolons is by searching annotations. We already do this in e.g. `p_rdrName`, you can check that out.

It looks like any of these options requires quite some work. I think we should delay supporting semicolons until after the first release. In the meantime let's get other stuff working well.

**Comment by mrkkrp** [comment]
I think it's worth dumping here results of a discussion we had. We go with option 3. That is, we'll try to insert semicolons where absolutely necessary to make output valid keeping its layout (single line/multiline). For example:

```haskell
f0 = let { a = 1; b = 2 } in b
```

Here the semicolon in necessary (inlike braces) so it'll be kept.

```haskell
foo = { do
      ; foo
      ; bar
      }
```

For something like this, `foo` and `bar` are on different lines. So the whole braces + semicolons thing is noise in this case, so we simplify it and end up with:

```haskell
foo = do
  foo
  bar
```

**Comment by utdemir** [comment]
Removing the braces is hard, I couldn't figure out a reliable and easy way to decide when we can omit the braces when printing.

One tricky thing to figure out what to do on single line-single statement layouts, since below two lines

```haskell
do { foo } `finally` bar
do foo `finally` bar
```

result in different ASTs. So we can not even remove braces even when there is a single statement (the current `master` also fails in this case).

The easiest workaround for now is to always use braces when using semicolon layout. Alternatively we can add a bunch of conditionals to OpApp to detect cases like this and force the braces.

**Comment by mrkkrp** [comment]
Resolved by #271.
